[#3186] Resource Header Vue Instance

Maurier opened the pull request:
This PR implements a Vue instance for the resource header as continuation of the work done in #3186. The following actions are now performed asynchronously and don't require a page refresh:
- Add Author
- Edit Author
- Delete Author
- Rearrange Authors

These changes can be tested at https://riverdev.hydroshare.org

Other changes
-------------------
- Style update to the Author type selection controls.
![image](https://user-images.githubusercontent.com/2448568/60736357-e663b580-9f13-11e9-8165-8177276dba89.png)
- Loading animation during ajax requests when rearranging Authors.
- Placeholder indicator when dragging an Author to rearrange.
![image](https://user-images.githubusercontent.com/2448568/60738012-01392880-9f1a-11e9-9422-591122567bde.png)

- HydroShare identifier input field has been replaced with a link to the HydroShare profile.
![image](https://user-images.githubusercontent.com/2448568/60737438-06957380-9f18-11e9-8d33-43d4700bee17.png)
- Minor layout adjustment to Delete Author confirmation
![image](https://user-images.githubusercontent.com/2448568/60737496-3b092f80-9f18-11e9-9637-df7ab97949c1.png)
- Responsive state handler for various buttons during ajax requests.



<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [ ] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [x] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [ ] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. Create a Resource
2. Add multiple authors.
3. Verify that editing/deleting/adding/rearranging authors is performed correctly and without page refresh.

cuahsimarko commented on hs_core/templates/pages/genericresource.html:
```diff
@@ -811,21 +817,17 @@ <h3>Resource Specific</h3>
     <script>
       {# GLOBAL DECLARATIONS FOR RESOURCE LANDING PAGE SCRIPTS #}
       const RES_KEYWORDS = {{ keywords|safe }};
       const SELF_ACCESS_LEVEL = "{{ self_access_level }}";
```
just so I understand here, the `|` was removed, correct? Does this have some significance for our version of Django?

Maurier replied:
Could you elaborate on this?

cuahsimarko replied:
Line 827 is showing pink, as though a character was removed:
![image](https://user-images.githubusercontent.com/45397838/61070635-2f68ad80-a3dd-11e9-9152-94652f96141b.png)

cuahsimarko commented on hs_core/templates/pages/genericresource.html:
```diff
@@ -835,6 +837,26 @@ <h3>Resource Specific</h3>
       const STATIC_URL = "{{ STATIC_URL }}";
       const USERS_JSON = {{ users_json|safe }};
+      const AUTHORS = [
+          {% for author in creators %}
```
Is it any cleaner to perform this in the python file that passes `creators` to the html template?

Maurier replied:
It's technically being done that way now, is just that the script tag exists inside the HTML file. But passing these data directly in the HTML would involve creating hidden components to attach this to. That would be way more inefficient and messy.

cuahsimarko commented on hs_core/views/__init__.py:
```diff
@@ -611,6 +611,17 @@ def delete_metadata_element(request, shortkey, element_name, element_id, *args,
+        ajax_response_data = {'status': 'success', 'message': "Author was deleted successfully"}
+    except Error as exp:
+        ajax_response_data = {'status': 'error', 'message': exp.message}
+    return JsonResponse(ajax_response_data)
```
optional - consistency on single vs double quotes

cuahsimarko replied:
does .status map to 200 / 404, 500 etc? I'm guessing you're following existing patterns in HydroShare but might be an opportunity to fix something if this is non-standard. https://docs.djangoproject.com/en/1.11/ref/request-response/#httpresponse-subclasses

cuahsimarko replied:
Implications across the codebase might be too large, anyway. Maybe just open a GH Issue if you find this is a bad practice across the repo

Maurier replied:
This property does not map to response status. It is simply used in the frontend to assert if the request was performed successfully or not.

cuahsimarko commented on theme/static/css/hydroshare_core.css:
```diff
@@ -399,6 +399,13 @@ a.sign-out-inline:hover {
+	overflow: hidden;
+}
+
 /* warning access removal dialog title */
```
does this belong in hydroshare_core.css or a smaller, more specific area?

cuahsimarko replied:
Looks like the other author stuff is already in there, so all set...

Maurier replied:
Correct. Right now there is not a more specific stylesheet to put this in, but that could change in the future if we decide to optimize this file.

cuahsimarko commented on theme/static/css/hydroshare_core.css:
```diff
@@ -983,33 +990,38 @@ li.account p{
 }
 
-.authors-wrapper > a:not(:first-child)::before, .authors-wrapper span:not(:first-child):not(.ui-sortable-helper) a::before {
+.authors-wrapper > a:not(:first-child)::before,
```
is the use of `child` selectors robust enough? Absent specific ids, always good to make sure it's not fragile and will withstand future refactoring.

Maurier replied:
Yes, this works perfectly for what it's trying to do. This adds a separator before each element and thus the first one doesn't need it because there is none before.

cuahsimarko commented on theme/static/css/hydroshare_core.css:
```diff
@@ -983,33 +990,38 @@ li.account p{
+	margin-left: 8px;
+	margin-right: 4px;
+	padding: 0;
+}
```
It's been many years, since I coded CSS in depth, so I was looking around and came across this guide. If you agree with its advice, maybe check out the margin section https://gist.github.com/basham/2175a16ab7c60ce8e001

Maurier replied:
In this case this component needs the units in 'em' because it is specifically designed to accommodate this element's font-size as the height. It can be reused like this anywhere on the site and it would keep using the correct value since 'em' refers to the font-size of the element.

cuahsimarko commented on theme/static/css/hydroshare_core.css:
```diff
@@ -3438,6 +3450,15 @@ span.you-flag {
 
+.fade-enter-active, .fade-leave-active {
+	transition: opacity .35s;
+	transition-timing-function: ease;
```
Is .35s a setting we should const/standardize/parameterize across similar elements ?

Maurier replied:
This is the animation time that bootstrap uses for collapse transition. This animation was put in place to replace the previous bootstrap collapse because it was not working inside the Vue instance and I could not figure out why. So, to keep it a little consistent I used the same time.

cuahsimarko commented on theme/static/js/generic-resource.js:
```diff
@@ -265,6 +265,14 @@ function update_download_status(task_id, download_path) {
 
+function getElementIndex(node) {
+    var index = 0;
+    while ((node = node.previousElementSibling)) {
```
consider let

cuahsimarko replied:
can you use .filter and a count?

cuahsimarko replied:
double equals? `==`

Maurier replied:
> can you use .filter and a count?
Not in this case, because I need a function that takes in an html element and returns its position index inside within its parent. These items could even look identical since two authors could have the same name, and thus there is not a clear way to filter.

Maurier replied:
> double equals? `==`

The assignment operation is intentional in this case.

Maurier replied:
> can you use .filter and a count?

After checking the code again, there is an attribute I added later which I could use to get this done through a filter function.

cuahsimarko commented on theme/static/js/resourceLandingAjax.js:
```diff
@@ -417,10 +417,9 @@ function remove_aggregation_ajax_submit(url) {
     return $.ajax({
         type: "POST",
-        url: '/hsapi/_internal/' + res_id + '/get-metadata/',
+        url: '/hsapi/_internal/' + SHORT_ID + '/get-metadata/',
```
is this a CONSTANT or variable?

Maurier replied:
This is a constant which is meant to replace all occurrences of res_id. I'll make sure to update these occurrences.

cuahsimarko commented on theme/static/js/resourceLandingAjax.js:
```diff
@@ -1561,69 +1560,23 @@ function setFileTypeMetadataFormsClickHandlers(){
-        }
-    }
+function updateResourceKeywords(keywords) {
+    subjKeywordsApp.$data.resKeywords = keywords;
```
can you just bind the model?

Maurier replied:
Agree. Will make this change.

cuahsimarko commented on theme/templates/includes/author-preview.html:
```diff
@@ -1,85 +1,92 @@
-                                    <a href="" data-researchgateid class="identifier-icon" target="_blank" data-toggle="tooltip" data-placement="bottom"
-                                       title="ResearchGate">
-                                        <i class="ai ai-researchgate-square hover-shadow"></i>
+<script id="author-preview-modal-template" type="text/x-template">
```
does it make sense to have the modal in a separate file or is that too extreme? I've been trying to break up colossal / run-on files and make top-level templates more readable. Downside, is tracing involves opening all the separate files... we should start to agree on some instincts of number of lines that warrants breaking things apart (1,000? 2,000?)

Maurier replied:
Indeed. We have made progress in separating the code a little better. We used to have these modals inside genericresource.html, and we moved it into modals.html not long ago. We would need to discuss if it is worth it to separate this HTML even further into files for each element.

Maurier commented:
This has been deployed to https://riverdev.hydroshare.org for testing purposes.

horsburgh commented:
@Maurier - I noticed a few issues:

* When an organization is selected, there should be no ability to enter author identifiers. Those are shown now on the form when Organization is selected, but shouldn't be.
* Need to make sure there is example text in the author name field indicating that the name should be entered with the last name first - e.g., "Horsburgh, Jeffery S." Otherwise, your recent work to sort out author naming issues will be for not.
* The modal window for entering contributor information should be consistent with this form. It still uses the radio buttons. Also, our text label on the contributors field says "People or organizations that contributed..." - but we only allow entry of people as contributors and not organizations. Should we be using the exact same form for both? If we allow organizations as contributors, there may be some back-end issues.
* The delete button for an existing user once they are selected on the form is not exactly aligned in the input area:
![image](https://user-images.githubusercontent.com/3485902/61238157-55ea5980-a6f9-11e9-883b-fcdbee57e39c.png)

These are the issues I found in doing a bit of testing just now in Chrome on my Mac.

Maurier commented:
@horsburgh - I think we should aim any changes in the Add Contributor workflow to be done inside the Vue instance I will be writing for this component in the near future. 

In the meantime, I have implemented the changes listed above for consistency. In the case of Add Contributor, I've added the same control with only the two options supported right now:
![image](https://user-images.githubusercontent.com/2448568/61395933-d2a93f00-a883-11e9-949e-6064c907a8f3.png)


![image](https://user-images.githubusercontent.com/2448568/61396124-3cc1e400-a884-11e9-83d5-772267cc862a.png)

Let me know if the wording should be changed to the text label above the contributors listing so that it's clear that Organizations can not be listed as Contributor (for now).

horsburgh commented:
@Maurier - I think these changes are fine. For now, since you cannot add an organization as a contributor, I would change the wording on the edit page to:

>People that contributed technically, materially, financially, or provided general support for the creation of the resource's content but are not considered authors.

sblack-usu commented on hs_core/views/__init__.py:
```diff
@@ -619,6 +619,17 @@ def delete_metadata_element(request, shortkey, element_name, element_id, *args,
+        res.metadata.delete_element('creator', element_id)
+        resource_modified(res, request.user, overwrite_bag=False)
+        ajax_response_data = {'status': 'success', 'message': "Author was deleted successfully"}
+    except Error as exp:
```
With try/catch it's best to be specific.  Keep code in the try block small and be specific about the error.  What errors are you trying to catch here?

Maurier replied:
Understood. In this case I was trying to catch database errors.

sblack-usu commented on hs_core/views/__init__.py:
```diff
@@ -619,6 +619,17 @@ def delete_metadata_element(request, shortkey, element_name, element_id, *args,
 
+def delete_author(request, shortkey, element_id, *args, **kwargs):
+    try:
+        res, _, _ = authorize(request, shortkey, needed_permission=ACTION_TO_AUTHORIZE.EDIT_RESOURCE)
```
authorize will raise NotFound and PermissionDenied, pull this out of the try block

sblack-usu commented on hs_core/views/__init__.py:
```diff
@@ -619,6 +619,17 @@ def delete_metadata_element(request, shortkey, element_name, element_id, *args,
+def delete_author(request, shortkey, element_id, *args, **kwargs):
+    try:
+        res, _, _ = authorize(request, shortkey, needed_permission=ACTION_TO_AUTHORIZE.EDIT_RESOURCE)
+        res.metadata.delete_element('creator', element_id)
```
delete_element will raise ValidationError if the creator element does not exist... I don't think there is any reason for this to be wrapped in the try/catch block.

sblack-usu replied:
only wrap it if you don't like the message passed up and want to add/modify it.  Be specific about what you want to catch.

Maurier replied:
@sblack-usu - In any case, a JSON response needs to be returned with `'status': 'error'` since this is expected by the frontend logic. Don't we need this fallback or is ValidationError the only exception that will be raised by this method?

cuahsimarko commented:
@sblack-usu @Maurier is this ready to merge? I'd like to pull it into my PR and start conflict resolution.

Maurier commented:
@cuahsimarko It is, just need re-approval. I had to solve some merge conflicts and do some code cleanup after it was approved.
